### PR TITLE
Add examples for HF datasets

### LINKED ISSUE #1352: [WIP] Examples for demonstrating the usage and incremental value of TorchData Nodes
### 🚀 The feature

Starting this issue to track minimal examples we can create to demonstrate effective usage and value of TorchData nodes. I can create separate issues for each of these as required.

### Motivation, pitch

1. Vanilla torch.utils dataloader usage ported over to torchdata nodes
2. GPU accelerated transforms
3. Flexible parallelism (mixing multiprocessing with multithreading)
4. Examples porting over popular OSS datasets
    -  connecting to popular cloud storage
5. Example creating new nodes (might get covered through examples above)
6. Basic multimodal model trained E2E using torchdata nodes
7. Chaining multiple transforms (might get covered through examples above)
8. Dataset mixing (with different sampling strategies)

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ramanishsingh**
This PR aims to add examples on how to use torchdata.nodes for dataloading and training models.
These examples aim to cover:
1. Load and process datasets from HuggingFace.
2. Using the functionalities of `nodes`: batching, loading, mapping
3. Getting batches and training the ML model.
4. Specific examples include image recognition (MNIST digit recognition), NLP (movie review sentiment classification using BERT), and usage of `MultiNodeWeightedSampler` in `nodes` to show the tools available in `nodes`.

Fixes #1352

### CODE REVIEW

File: examples/nodes/hf_datsets.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+from torchdata.nodes.adapters import IterableWrapper
+from torchdata.nodes.batch import Batcher
+from torchdata.nodes.map import Mapper, ParallelMapper
+from torchdata.nodes.prefetch import Prefetcher
+
+ds = Dataset.from_dict({"data": data})
+
+source_node = IterableWrapper(ds)
+
+print("==" * 20)
```
Comment (divyanshk) [inline examples/nodes/hf_datsets.py:14]: One thing I had in mind with /examples, we don't expect users to really run the code, but mostly use it as a reference implementation to port over their workflows. I wonder if we can write this file such that users can see all the relevant nodes linked together clearly?

File: examples/nodes/hf_datsets.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+from datasets import Dataset  # make sure that you have huggingface datasets installed
+
+data = [[1, 2], [3, 4], [5, 6], [7, 8], [9, 10]]
```
[the remark this answers is not in the document]
Comment (ramanishsingh) [reply examples/nodes/hf_datsets.py:3]: Yeah, that's my goal as well. 
I am facing some fwdproxy issues with loading the datasets from huggingface.
I am trying to resolve that, after that I can include examples with more complex datasets.

File: examples/nodes/hf_datsets.py (hunk 1 above)
Comment (ramanishsingh) [reply examples/nodes/hf_datsets.py:14]: Yes, sure.
I will port these examples to a Jupyter notebook with print outputs in the file so it is easier to get the outputs.
Will also add some interleaved examples where we use multiple node functionalities together.

**Comment by divyanshk** [comment]
@ramanishsingh Can you add pointers to what all these three notebooks aim to cover in the PR top comment?

File: examples/nodes/utils.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+import torch
+from torch.utils.data import default_collate
+
+
+def map_fn_bert(item, max_len, tokenizer):
```
Comment (andrewkho) [inline examples/nodes/utils.py:5]: This map_fn should be in the notebook

File: examples/nodes/utils.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        data = []
+        for line_number, line in enumerate(csvfile):
+            if line_number == 0:
+                continue
+            row = [float(x) for x in line.strip().split(",")]
+            data.append(row)
+    return data
+
+
+class MDS_Net(nn.Module):
```
Comment (divyanshk) [inline examples/nodes/utils.py:256]: The model like a simple MLP. What does MDS stand for? multidataset?

Comment (ramanishsingh) [reply examples/nodes/utils.py:256]: Yes, but if it is confusing I can modify the name to something simpler.

File: examples/nodes/mds_data/gaussian_dataset.csv [hunk 5]
```diff
@@ -0,0 +1,2001 @@
+x,y
```
Comment (andrewkho) [inline examples/nodes/mds_data/gaussian_dataset.csv:1]: Can you just call this folder "data" or "example_data"?

File: examples/nodes/utils.py (hunk 3 above)
Comment (andrewkho) [reply examples/nodes/utils.py:12]: Not sure if I'm looking at something old, but should we delete this function?

File: examples/nodes/utils.py (hunk 4 above)
Comment (andrewkho) [reply examples/nodes/utils.py:256]: Just call this MLP

File: examples/nodes/utils.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+        self.fc3 = nn.Linear(128, 1)  # hidden layer (128) -> output layer (1)
+
+    def forward(self, x):
+        x = self.relu(self.fc1(x))
+        x = self.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+
+
+def train_mds_model(datasets, weights, batch_size=512):
```
Comment (andrewkho) [inline examples/nodes/utils.py:272]: can we have a single "train" function that works for all 3 examples? They all look somewhat similar

**Comment by andrewkho** [comment]
Not sure how to comment on notebooks so will add comments here:
image
do you need to a 4-tuple here, can it just be a 2-tuple?

**Comment by andrewkho** [comment]
nit: can you rename `multiple_datasets` to `multidatasets`?

File: examples/nodes/hf_datasets_nodes_mnist.ipynb [hunk 7]
```diff
@@ -0,0 +1,172 @@
+{
+ "cells": [
```
Comment (andrewkho) [inline examples/nodes/hf_datasets_nodes_mnist.ipynb:2]: Add a title

File: examples/nodes/hf_datasets_nodes_mnist.ipynb [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    "    return {\"image\":image, \"label\":label}\n",
+    "    \n",
+    "node = ParallelMapper(node, map_fn=map_fn, num_workers=2) # output items are Dict[str, tensor]\n",
+    "\n",
+    "\n",
+    "# Hyperparameters\n",
+    "batch_size = 2 #batch size is kept low so that we can easily see the bacthes when we print them in the later cells\n",
+    "\n",
+    "# Next we batch the inputs, and then apply a collate_fn with another Mapper\n",
+    "# to stack the tensors between. We use torch.utils.data.default_collate for this\n",
```
Comment (andrewkho) [inline examples/nodes/hf_datasets_nodes_mnist.ipynb:72]:
```suggestion
    "# to stack the tensor. We use torch.utils.data.default_collate for this\n",
```

File: examples/nodes/hf_datasets_nodes_mnist.ipynb [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    "node = Batcher(node, batch_size=batch_size) # output items are List[Dict[str, tensor]]\n",
+    "node = ParallelMapper(node, map_fn=default_collate, num_workers=2) # outputs are Dict[str, tensor]\n",
+    "\n",
+    "# we can optionally apply pin_memory to the batches\n",
+    "if torch.cuda.is_available():\n",
+    "    node = PinMemory(node)\n",
+    "\n",
+    "# Since nodes are iterators, they need to be manually .reset() between epochs.\n",
+    "# We can wrap the root node in Loader to convert it to a more conventional Iterable.\n",
```
Comment (andrewkho) [inline examples/nodes/hf_datasets_nodes_mnist.ipynb:81]:
```suggestion
    "# Instead, we can wrap the root node in Loader to convert it to a more conventional Iterable.\n",
```

File: examples/nodes/hf_datasets_nodes_mnist.ipynb [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+      "text/plain": [
+       "<Figure size 800x600 with 2 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# Once we have the loader, we can get batches from it over multiple epochs, to train the ML model\n",
```
Comment (andrewkho) [inline examples/nodes/hf_datasets_nodes_mnist.ipynb:126]:
```suggestion
    "# Once we have the loader, we can get batches from it over multiple epochs, to train the model\n",
```

File: examples/nodes/hf_imdb_bert.ipynb [hunk 11]
```diff
@@ -0,0 +1,148 @@
+{
```
Comment (andrewkho) [inline examples/nodes/hf_imdb_bert.ipynb:1]: Add a title

File: examples/nodes/hf_imdb_bert.ipynb [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+   "outputs": [],
+   "source": [
+    "from torchdata.nodes import MapStyleWrapper, ParallelMapper, Batcher, PinMemory, Loader\n",
+    "node = MapStyleWrapper(map_dataset=dataset, sampler=sampler)\n",
+    "\n",
+    "# Now we want to transform the raw inputs. We can just use another Mapper with\n",
+    "# a custom map_fn to perform this. Using ParallelMapper allows us to use multiple\n",
+    "# threads (or processes) to parallelize this work and have it run in the background\n",
+    "max_len = 512\n",
+    "batch_size = 2 # Keeping batch size smaller to easily inspect the outputs of a batch\n",
```
Comment (andrewkho) [inline examples/nodes/hf_imdb_bert.ipynb:70]:
```suggestion
    "batch_size = 2\n",
```

**Review by andrewkho — CHANGES_REQUESTED** [review]
Few nits but otherwise looks good

**Review by andrewkho — APPROVED** [review]

File: examples/nodes/hf_imdb_bert.ipynb [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+   "execution_count": 3,
+   "id": "2afac7d9-3d66-4195-8647-dc7034d306f2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Load IMDB dataset from huggingface datasets and select the \"train\" split\n",
+    "dataset = load_dataset(\"imdb\", streaming=False)\n",
+    "dataset = dataset[\"train\"]\n",
+    "# Since dataset is a Map-style dataset, we can setup a sampler to shuffle the data\n",
+    "# Please refer to the migration guide here https://pytorch.org/data/docs/build/html/migrate_to_nodes_from_utils.html\n",
```
Comment (andrewkho) [inline examples/nodes/hf_imdb_bert.ipynb:46]: I gave you the wrong link, after nightly run this link is live now: https://pytorch.org/data/main/migrate_to_nodes_from_utils.html

**Review by andrewkho — APPROVED** [review]
update URL and then good to go

**Pull request merged**